src/java/org/apache/cassandra/db/Directories.java

```diff
@@ -274,6 +294,76 @@ public boolean accept(File file)
         }
     }
 
+    public static void scheduleVerifyingDiskDoesNotExceedThresholdChecks()
+    {
+        ScheduledExecutors.scheduledTasks.scheduleAtFixedRate(
+                                getVerifyDiskHasEnoughUsableSpaceRunnable(), 1, 1, TimeUnit.MINUTES);
+    }
+
+    @VisibleForTesting
+    static Runnable getVerifyDiskHasEnoughUsableSpaceRunnable()
+    {
+        return () -> {
+            try {
+                Directories.verifyDiskHasEnoughUsableSpace();
+            } catch (FSWriteError e) {
+                // If node is already disabled propagating the exception from scheduled executor is redundant and clogs logs.
+                if (!StorageService.instance.isNodeDisabled()) {
+                    throw e;
+                }
+            }
+        };
+    }
+
+    public static void verifyDiskHasEnoughUsableSpace()
+    {
+        Map<DataDirectory, Double> dirsToUtilization = Arrays.stream(dataDirectories)
+                                                             .collect(Collectors.toMap(
+                                                             Function.identity(),
+                                                             dir -> {
+                                                                 long total = dir.getTotalSpace();
+                                                                 long used = total - dir.getAvailableSpace();
+                                                                 return (double) used / total;
+                                                             })
+                                                             );
+
+        Map.Entry<DataDirectory, Double> maxPathToUtilization = dirsToUtilization
+                                                                .entrySet()
+                                                                .stream()
+                                                                .max(Map.Entry.comparingByValue())
+                                                                .orElseThrow(() -> new RuntimeException(
+                                                                "Failed to filter most full data directory"));
+        logger.debug(String.format("Highest data directory disk utilization on path %s (%f)",
+                                   maxPathToUtilization.getKey().location, maxPathToUtilization.getValue()));
+        double threshold = DatabaseDescriptor.getMaxDiskUtilizationThreshold();
+        double currUtilization = maxPathToUtilization.getValue();
+        if (currUtilization >= threshold)
+        {
+            ExceededDiskThresholdException cause = new ExceededDiskThresholdException(
+                                    maxPathToUtilization.getKey().location, currUtilization, threshold);
+            throw new FSWriteError(cause, maxPathToUtilization.getKey().location);
```

Comment:
Why did we decide to wrap this in an FSWriteError?  We aren't actually attempting to write anything at the time of the failure.

Reply:
Hm, I think my logic at the start was since it will be handled by the `DefaultFSErrorHandler` it should have been wrapped with a subclass of `FSError`, so either `FSWriteError` or `FSReadError`. Looking back at it now though I agree it makes sense for it to remain independent.